File: src/point.cpp

#include <math.h>
#include <sstream>
#include <iomanip>
#include "util.hpp"
#include "point.hpp"

Point::Point(double x, double y) {
    this->x = x;
    this->y = y;
    this->z = 0;
}

Point::Point(double x, double y, double z) {
    this->x = x;
    this->y = y;
    this->z = z;
}

Point::~Point() {

}

double Point::get_x() const {
    return this->x;
}

double Point::get_y() const {
    return this->y;
}

double Point::get_z() const {
    return this->z;
}

void Point::set_x(double x) {
    this->x = x;
}

void Point::set_y(double y) {
    this->y = y;
}

void Point::set_z(double z) {
    this->z = z;
}

double Point::euclid_dist(Point other_point) {

    double px = pow(this->x - other_point.get_x(), 2);
    double py = pow(this->y - other_point.get_y(), 2);
    double pz = pow(this->z - other_point.get_z(), 2);
    return sqrt(px + py + pz);

}

Point Point::to_unit_vector() {
    Point origin(0, 0, 0);
    double mag = this->euclid_dist(origin);
    Point p(this->x / mag, this->y / mag, this->z / mag);
    return p;
}

std::string Point::str() {
    int precision = 4;
    std::stringstream buffer;
    buffer << std::setprecision(precision) << this->x
        << " " << std::setprecision(precision) << this->y
        << " " << std::setprecision(precision) << this->z;
    return buffer.str();
}

bool Point::operator==(Point rhs) {
    return this->str().compare(rhs.str()) == 0;
}

Point Point::get_random_point_2d(double width, double height) {
    double x = rand_float(0, width);
    double y = rand_float(0, height);
    Point p(x, y);
    return p;
}

File: scripts/point.py

import math
import stpoint
import random


class Point(object):

    def __init__(self, x, y, z=0):
        self.x = x
        self.y = y
        self.z = z

    def get_x(self):
        return self.x

    def get_y(self):
        return self.y

    def get_z(self):
        return self.z

    def set_x(self, x):
        self.x = x
        return self

    def set_y(self, y):
        self.y = y
        return self

    def set_z(self, z):
        self.z = z
        return self

    def dist_to(self, other_point):
        return math.sqrt(
            pow(self.x - other_point.x, 2) +
            pow(self.y - other_point.y, 2) +
            pow(self.z - other_point.z, 2)
        )

    def to_unit_vector(self):
        mag = self.dist_to(Point(0, 0, 0))

        if mag == 0:
            return Point(0, 0, 0)
        else:
            return Point(self.x / mag, self.y / mag, self.z / mag)

    def to_st_point_2d(self, t):
        return stpoint.make(self.x, self.y, t)

    def to_list(self):
        return [self.x, self.y, self.z]

    def __str__(self):
        return "X: {0}, Y: {1}, Z: {2}".format(self.x, self.y, self.z)

    def __repr__(self):
        return "Point({0}, {1}, {2})".format(self.x, self.y, self.z)

    def __hash__(self):
        return hash(str(self))

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y, self.z + other.z)

    def __mult__(self, scalar):
        return Point(scalar * self.x, scalar * self.y, scalar * self.z)

    def __eq__(self, val):
        try:
            return val.x == self.x and val.y == self.y and val.z == self.z
        except:
            return False


def make(*args, **kwargs):
    return Point(*args, **kwargs)


def get_random_point(width, height):
    x = random.uniform(0, width)
    y = random.uniform(0, height)

    return Point(x, y)


def get_random_point_3d(width, height, altitude):
    p = get_random_point(width, height)
    p.set_z(random.randint(0, altitude))

    return p
